File: 456.132-pattern.java
import java.util.Stack;

class Solution {
	public boolean find132pattern(int[] nums) {
		for (int n = nums.length, i = n - 1, top = n, third = Integer.MIN_VALUE; i >= 0; i--) {
			if (nums[i] < third)
				return true;
			while (top < n && nums[i] > nums[top])
				third = nums[top++];
			nums[--top] = nums[i];
		}

		return false;
	}
}

// bool find132pattern(vector<int>& nums) {
//     int s3 = INT_MIN;
//     stack<int> st;
//     for( int i = nums.size()-1; i >= 0; i -- ){
//         if( nums[i] < s3 ) return true;
//         else while( !st.empty() && nums[i] > st.top() ){ 
//           s3 = st.top(); st.pop(); 
//         }
//         st.push(nums[i]);
//     }
//     return false;
// }

File: 456.132-pattern.js
/*
 * @lc app=leetcode id=456 lang=javascript
 *
 * [456] 132 Pattern
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var find132pattern = function (nums) {
  const stack = [];
  const n = nums.length;
  let k = Number.NEGATIVE_INFINITY;

  for (let i = n - 1; i >= 0; --i) {
    if (nums[i] < k) {
      return true;
    }
    while (stack && stack[stack.length - 1] < nums[i]) {
      k = Math.max(k, stack.pop());
    }
    stack.push(nums[i]);
  }

  return false;
};
// @lc code=end
